File: Go/fibonacci.go
/*
 * In The Name Of God
 * ========================================
 * [] File Name : fibonacci.go
 *
 * [] Creation Date : 28-05-2015
 *
 * [] Created By : Parham Alvani (parham.alvani@gmail.com)
 * =======================================
 */
/*
 * Copyright (c) 2015 Parham Alvani.
 */

package fibonacci

import "math"

// Fibonacci implementation constants
const (
	Recursive = iota
	Memoized
	Linear
	Matrix
	FastDoubling
	ClosedForm
)

// Implementation provides abstraction for fibonacci implmeentations
type Implementation interface {
	Fibonacci(int) int
}

// New returns asked implementation of fibonacci
func New(t int) Implementation {
	switch t {
	case Recursive:
		return new(recursive)
	case Memoized:
		return newMemoized()
	case Linear:
		return new(linear)
	case Matrix:
		return new(matrix)
	case FastDoubling:
		return new(fastDoubling)
	case ClosedForm:
		return new(closedForm)
	default:
		return nil
	}
}

type recursive struct{}

func (r *recursive) Fibonacci(n int) int {
	if n == 0 {
		return 0
	}

	if n == 1 {
		return 1
	}

	return r.Fibonacci(n-1) + r.Fibonacci(n-2)
}

// memoized is the recursive implementation with each F(n) cached the first
// time it is computed, which collapses the exponential call tree to O(n).
// It carries state, so unlike the others an instance is worth reusing.
type memoized struct {
	cache map[int]int
}

func newMemoized() *memoized {
	return &memoized{cache: map[int]int{0: 0, 1: 1}}
}

func (m *memoized) Fibonacci(n int) int {
	if f, ok := m.cache[n]; ok {
		return f
	}

	f := m.Fibonacci(n-1) + m.Fibonacci(n-2)
	m.cache[n] = f

	return f
}

type linear struct{}

func (r *linear) Fibonacci(n int) int {
	a := 0
	b := 1

	for range n {
		a, b = b, a+b
	}

	return a
}

// mat2 is a 2x2 matrix laid out in row-major order.
type mat2 [2][2]int

func (m mat2) mul(o mat2) mat2 {
	return mat2{
		{m[0][0]*o[0][0] + m[0][1]*o[1][0], m[0][0]*o[0][1] + m[0][1]*o[1][1]},
		{m[1][0]*o[0][0] + m[1][1]*o[1][0], m[1][0]*o[0][1] + m[1][1]*o[1][1]},
	}
}

// matrix raises [[1,1],[1,0]] to the n'th power, which puts F(n) in the
// top-right cell, using exponentiation by squaring to do it in O(log n).
type matrix struct{}

func (m *matrix) Fibonacci(n int) int {
	result := mat2{{1, 0}, {0, 1}}
	base := mat2{{1, 1}, {1, 0}}

	for e := n; e > 0; e >>= 1 {
		if e&1 == 1 {
			result = result.mul(base)
		}

		base = base.mul(base)
	}

	return result[0][1]
}

// fastDoubling derives F(2k) and F(2k+1) from F(k) and F(k+1) directly:
//
//	F(2k)   = F(k) * (2*F(k+1) - F(k))
//	F(2k+1) = F(k)^2 + F(k+1)^2
//
// This is the matrix method with the matrix algebra folded out by hand, so it
// is also O(log n) but without the redundant multiplications.
type fastDoubling struct{}

func (f *fastDoubling) Fibonacci(n int) int {
	a, _ := f.pair(n)

	return a
}

// pair returns F(n) and F(n+1) together, halving n on each step.
func (f *fastDoubling) pair(n int) (int, int) {
	if n == 0 {
		return 0, 1
	}

	a, b := f.pair(n / 2)
	c := a * (2*b - a)
	d := a*a + b*b

	if n%2 == 0 {
		return c, d
	}

	return d, c + d
}

// closedForm evaluates Binet's formula. It is the only implementation here
// that does not compute F(n) exactly: it rounds a float64 to the nearest
// integer, and the rounding error in math.Pow compounds with n until it
// exceeds 1/2 and the rounding lands on the wrong integer. That happens at
// n=76, well before F(n) itself outgrows a float64's 53-bit mantissa, so the
// limit here is precision and not range. See TestClosedFormPrecisionLimit.
type closedForm struct{}

func (c *closedForm) Fibonacci(n int) int {
	sqrt5 := math.Sqrt(5)
	phi := (1 + sqrt5) / 2
	psi := (1 - sqrt5) / 2

	return int(math.Round((math.Pow(phi, float64(n)) - math.Pow(psi, float64(n))) / sqrt5))
}

File: C/fibonacci.c
/**
 * fibonacci.c
 *
 * The Fibonacci sequence by all 6 algorithms in C.
 */

#include "fibonacci.h"
#include "cc.h" // https://github.com/JacksonAllan/CC

#include <math.h>
#include <string.h>

/**
 * recursive
 */

// fibonacci_recursive is a recursive version of Fibonacci computation.
static int64 fibonacci_recursive(int n) {
  if (n < 0)
    return ERRNO;
  if (n == 0)
    return 0;
  if (n == 1)
    return 1;

  return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2);
}

/**
 * memoized
 */

// fibonacci_memoized is an improvement to fibonacci_recursive,
// using a hash map to cache computed Fibonacci values to avoid
// duplicated computing.
static int64 fibonacci_memoized_worker(map(int, int64) * cache, int n) {
  if (n < 2)
    return n;

  int64 *value = get(cache, n);
  if (value != NULL) {
    return *value;
  }

  int64 first = fibonacci_memoized_worker(cache, n - 1);
  int64 second = fibonacci_memoized_worker(cache, n - 2);
  if (first == ERRNO || second == ERRNO)
    return ERRNO;

  int64 ret = first + second;

  // cache computed value
  if (!insert(cache, n, ret)) {
    fprintf(stderr, "failed to insert into map\n");
    return ERRNO;
  }

  return ret;
}

static int64 fibonacci_memoized(int n) {
  if (n < 0)
    return ERRNO;

  // Using github.com/JacksonAllan/CC, where CC stands for Convenient
  // Containers. This single-header project provisions a set of modern hash map
  // API.
  map(int, int64) cache;
  init(&cache);

  int64 ret = fibonacci_memoized_worker(&cache, n);
  cleanup(&cache);
  return ret;
}

/**
 * linear
 */

// fibonacci_linear is an iteration version of Fibonacci computing algorithms.
// This version leverages two variables to store values needed for computation
// in a dynamic programming way, thus reducing space complexity to O(1).
static int64 fibonacci_linear(int n) {
  int64 last = 0, curr = 1;
  for (int i = 0; i < n; i++) {
    int64 temp = curr;
    curr += last;
    last = temp;
  }

  return last;
}

/**
 * matrix
 */

typedef struct {
  int64 data[2][2];
} mat2_t;

static mat2_t mul(mat2_t mat_A, mat2_t mat_B) {
  return (mat2_t){
      .data =
          {
              {mat_A.data[0][0] * mat_B.data[0][0] +
                   mat_A.data[0][1] * mat_B.data[1][0],
               mat_A.data[0][0] * mat_B.data[0][1] +
                   mat_A.data[0][1] * mat_B.data[1][1]},
              {mat_A.data[1][0] * mat_B.data[0][0] +
                   mat_A.data[1][1] * mat_B.data[1][0],
               mat_A.data[1][0] * mat_B.data[0][1] +
                   mat_A.data[1][1] * mat_B.data[1][1]},
          },
  };
}

// fibonacci_matrix raises the Fibonacci transformation matrix {{1, 1}, {1, 0}}
// to the n-th power using exponentiation by squaring. F(n) is stored in the
// top-right cell of the resulting matrix, giving O(log n) time complexity.
static int64 fibonacci_matrix(int n) {
  mat2_t result = {
      .data = {{1, 0}, {0, 1}},
  };
  mat2_t base = {
      .data = {{1, 1}, {1, 0}},
  };

  for (int e = n; e > 0; e >>= 1) {
    if ((e & 1) == 1) {
      result = mul(result, base);
    }
    base = mul(base, base);
  }
  return result.data[0][1];
}

/**
 * fast doubling
 */

typedef struct {
  int64 first, second;
} tuple;

// pair returns F(n) and F(n+1) together. It recursively computes the pair
// for n/2, then derives the requested pair with the fast-doubling identities.
static tuple pair(int n) {
  if (n == 0)
    return (tuple){0, 1};

  tuple half = pair(n / 2);
  int64 a = half.first, b = half.second;

  int64 c = a * (2 * b - a);
  int64 d = a * a + b * b;

  if (n % 2 == 0)
    return (tuple){c, d};

  return (tuple){d, c + d};
}

// fibonacci_fast_doubling computes Fibonacci numbers from the identities
// F(2k) = F(k) * (2 * F(k + 1) - F(k)) and
// F(2k + 1) = F(k)^2 + F(k + 1)^2, giving O(log n) time complexity.
static int64 fibonacci_fast_doubling(int n) { return pair(n).first; }

/**
 * closed form
 */

// fibonacci_closed_form evaluates Binet's formula using the golden ratio and
// its conjugate. The result is rounded to the nearest integer, but floating-
// point precision means this implementation is not exact for sufficiently
// large values of n.
static int64 fibonacci_closed_form(int n) {
  double phi = (1 + sqrt(5)) / 2;
  double psi = (1 - sqrt(5)) / 2;

  return (int64)(round((pow(phi, (double)n) - pow(psi, (double)n)) / sqrt(5)));
}

// fibonacci is the main entrance for all implementations.
int64 fibonacci(enum fib_algorithm algorithm, int n) {
  switch (algorithm) {
  case RECURSIVE:
    return fibonacci_recursive(n);
  case MEMOIZED:
    return fibonacci_memoized(n);
  case LINEAR:
    return fibonacci_linear(n);
  case MATRIX:
    return fibonacci_matrix(n);
  case FAST_DOUBLING:
    return fibonacci_fast_doubling(n);
  case CLOSED_FORM:
    return fibonacci_closed_form(n);
  default:
    return ERRNO;
  }
}
